rfc/001-truncated-block-history.md

```diff
@@ -0,0 +1,87 @@
+
+* Snapshot retention (SDK): the number of recent state sync snapshots to retain.
+
+## Proposal
+
+* Unbonding time (SDK): already implemented as a genesis parameter `app_state.staking.params.unbonding_time`. Must now satisfy `unbonding_time < 2 * block.retention * block.time_iota_ms` otherwise errors (factor of 2 since `block.time_iota_ms` does not take into account block execution time).
+
+* Block time (Tendermint): already implemented as a genesis parameter `consensus_params.block.time_iota_ms`.
+
+* Block retention (Tendermint): implement as a new genesis parameter `consensus_params.block.retention` (default 0, i.e. retain all), and add a local configuration option `consensus.prune_blocks` (default off) that allows node operators to either enable or disable block pruning on a per-node basis. Must be at minimum 2, since the current and previous blocks are required for progress.
+
+* Snapshot interval (SDK): implement as a new genesis parameter `app_state.snapshots.interval` (default 0, i.e. disabled). Must satisfy `interval < 2 * block.retention` otherwise errors, to allow state synced nodes to catch up via block replay (factor of 2 to account for time spent taking a snapshot).
+
+* Snapshot retention (SDK): add a local configuration option `snapshot-enable` (default off) to enable/disable snapshots on a per-node basis, and `snapshot-retention` (default 2, 0 retains all) that specifies the number of snapshots to keep. Must be at least 2 to allow in-flight state syncs to complete as new snapshots are produced, but higher retention can be useful e.g. during mixed-version upgrades where old snapshots can be used by old-version nodes, or for state machine archives.
+
+## Status
+
+Proposed
+
+## Consequences
+
+### Positive
+
+* Genesis parameters for block retention, block time, and unbonding time allows governance consensus to ensure availability and correctness of light client verification and short-term chain auditability, with enforced sanity checks.
+
+* Genesis parameters for snapshot interval ensures uniform snapshot schedules across nodes, providing high availability of snapshots at a given height for parallel downloads and faster state syncing, with enforced sanity checks.
+
+* Node operators can independently decide whether they want to provide complete block histories and snapshots.
+
+### Negative
+
+* Social coordination is required to run archival nodes, failure to do so may lead to permanent loss of historical blocks.
+
+* Social coordination is required to run snapshot nodes, failure to do so may lead to inability to run state sync, and inability to bootstrap new nodes at all if no archival nodes are online.
+
+* Genesis parameter can only be changed by creating a new chain. This means that enabling block pruning or state sync, as well as modifying the settings, would be a non-trivial operation.
+
+* Chain-wide block retention sets a lower bound on disk space requirements for all nodes.
```

Comment:
This is negative?

Reply:
If it was not a chain-wide consensus parameter then node operators would be free to choose even lower retention. In principle, it would be possible to run a node which did not retain blocks at all, only state.